walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,74 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 import java.util.Collection;
+import java.util.Collections;
+
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return Collections.emptyList();
+        }
+
+        List<TimeRange> conflictTimes = new ArrayList<TimeRange>();
+        List<TimeRange> conflictTimesIncludingOptional = new ArrayList<TimeRange>();
+
+        for (Event event : events) {
+            if (!Collections.disjoint(request.getAttendees(), event.getAttendees())) {
+                conflictTimes.add(event.getWhen());
+                conflictTimesIncludingOptional.add(event.getWhen());
+            }
+
+            if (!Collections.disjoint(request.getOptionalAttendees(), event.getAttendees())) {
+                conflictTimesIncludingOptional.add(event.getWhen());
+            }
+        }
+
+        Collections.sort(conflictTimes, TimeRange.ORDER_BY_START);
+        Collections.sort(conflictTimesIncludingOptional, TimeRange.ORDER_BY_START);
+
+        List<TimeRange> meetingTimes = findAvailableTimes(conflictTimesIncludingOptional, request.getDuration());
+        if (meetingTimes.isEmpty()) {
+            if (request.getAttendees().isEmpty() && !request.getOptionalAttendees().isEmpty()) {
+                return meetingTimes;
```

Comment:
prefer to arrange this code so that there is only 1 return statement. this makes the code simpler to reason about, rather than having multiple return points scattered through out the function. a common exception to this rule is early returns for error conditions or shortcuts (like you have at the top of this function).

Reply:
Fixed. Now there is only one return statement and one if statement.